File: data-code/_HCRIS_Data.R
# Meta --------------------------------------------------------------------

## Author:        Ian McCarthy
## Date Created:  5/30/2019
## Date Edited:   1/7/2026
## Notes:         -- v1996 of the HCRIS forms run through 2011 due to lags 
##                   in processing and hospital fiscal years


# Preliminaries -----------------------------------------------------------
if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse, ggplot2, dplyr, lubridate)


# Import data -------------------------------------------------------------

## create v1996 and v2010 data
source('data-code/H1_HCRISv1996.R')
source('data-code/H2_HCRISv2010.R')

## create missing variables for columns introduced in v2010 of hcris forms
final.hcris.v1996 <- final.hcris.v1996 %>%
  mutate(hvbp_payment=NA, hrrp_payment=NA, tot_uncomp_care_charges=NA, tot_uncomp_care_partial_pmts=NA, bad_debt=NA,
         accum_dep = if_else(
           if_all(starts_with("depr_"), is.na),
           NA_real_,
           rowSums(across(starts_with("depr_"), ~ abs(.x)), na.rm = TRUE))) %>%
  select(-c(starts_with("depr_")))

## create missing variables for columns in v1996 that we have to compute in v2010
final.hcris.v2010 <- final.hcris.v2010 %>%
  mutate(uncomp_care=tot_uncomp_care_charges - tot_uncomp_care_partial_pmts + bad_debt,
         accum_dep = if_else(
           if_all(starts_with("depr_"), is.na),
           NA_real_,
           rowSums(across(starts_with("depr_"), ~ abs(.x)), na.rm = TRUE))) %>%
  select(-c(starts_with("depr_")))

## combine v1996 and v2010 hcris forms, and sort by provider_number/year
final.hcris=rbind(final.hcris.v1996,final.hcris.v2010) %>%
  mutate(fy_end=mdy(fy_end),fy_start=mdy(fy_start),
         date_processed=mdy(date_processed),date_created=mdy(date_created),
         tot_discounts=abs(tot_discounts), hrrp_payment=abs(hrrp_payment)) %>%
  mutate(fyear=year(fy_end)) %>%
  arrange(provider_number,fyear) %>%
  select(-year)

## count of hospitals/provider_number by year
final.hcris %>% group_by(fyear) %>% count()


# Clean data --------------------------------------------------------------

## create count of reports by hospital fiscal year
final.hcris <- 
  final.hcris %>% 
  add_count(provider_number, fyear, name="total_reports")

## create running total of reports
final.hcris <- 
  final.hcris %>% 
  group_by(provider_number, fyear) %>%
  mutate(report_number=row_number())

## identify hospitals with only one report per fiscal year 
## this will be the first set of hospitals in the final dataset
unique.hcris1 <- 
  final.hcris %>%
  filter(total_reports==1) %>%
  select(-report, -total_reports, -report_number, -npi, -status) %>%
  mutate(source='unique reports')


## identify hospitals with multiple reports per fiscal year
duplicate.hcris <- 
  final.hcris %>%
  filter(total_reports>1) %>%
  mutate(time_diff=fy_end-fy_start)

## calculate elapsed time between fy start and fy end for hospitals with multiple reports
duplicate.hcris <- 
  duplicate.hcris %>% 
  group_by(provider_number, fyear) %>%
  mutate(total_days=sum(time_diff))

## if the elapsed time within a fy sums to around 365, then just take the total of the two
## this will be the second set of hospitals in the final dataset
unique.hcris2 <- 
  duplicate.hcris %>%
  filter(total_days<370) %>%
  group_by(provider_number, fyear) %>%
  mutate(hrrp_payment=if_else(is.na(hrrp_payment),0,hrrp_payment),
         hvbp_payment=if_else(is.na(hvbp_payment),0,hvbp_payment)) %>%
  summarize(beds=max(beds), tot_charges=sum(tot_charges), tot_discounts=sum(tot_discounts),
            tot_operating_exp=sum(tot_operating_exp), ip_charges=sum(ip_charges),
            icu_charges=sum(icu_charges), ancillary_charges=sum(ancillary_charges),
            tot_discharges=sum(tot_discharges), mcare_discharges=sum(mcare_discharges),
            mcaid_discharges=sum(mcaid_discharges), tot_mcare_payment=sum(tot_mcare_payment),
            secondary_mcare_payment=sum(secondary_mcare_payment), hvbp_payment=sum(hvbp_payment),
            hrrp_payment=sum(hrrp_payment), fy_start=min(fy_start), fy_end=max(fy_end),
            date_processed=max(date_processed), date_created=min(date_created), 
            street=first(street), city=first(city), state=first(state),
            zip=first(zip), county=first(county), name=first(name),
            uncomp_care = sum(uncomp_care), tot_uncomp_care_charges=sum(tot_uncomp_care_charges),
            tot_uncomp_care_partial_pmts=sum(tot_uncomp_care_partial_pmts), 
            bad_debt=sum(bad_debt), cost_to_charge=first(cost_to_charge),
            new_cap_ass=sum(new_cap_ass), 
            cash=sum(cash),
            net_pat_rev=sum(net_pat_rev),
            fixed_assets=sum(fixed_assets),
            accum_dep=sum(accum_dep),
            current_assets=sum(current_assets),
            current_liabilities=sum(current_liabilities)
            ) %>%
  mutate(source='total for year')

## identify hospitals with more than one report and with elapsed time exceeding 370 days
duplicate.hcris2 <- 
  duplicate.hcris %>%
  filter(total_days>=370) %>%
  mutate(max_days=max(time_diff), max_date=max(fy_end))

## identify hospitals with one report (out of multiple) that appears to cover the full year
## this will be the third set of hospitals in the final dataset
unique.hcris3 <- 
  duplicate.hcris2 %>%
  filter(max_days==time_diff, time_diff>360, max_date==fy_end) %>%
  select(-report, -total_reports, -report_number, -npi, -status, -max_days, -time_diff, -total_days, -max_date) %>%
  mutate(source='primary report')

## identify remaining hospitals (those with more than one report that cover more than one full year and that do
##   not appear to have one report that takes up the full year)
## these hospitals appear to have changed their fiscal years
duplicate.hcris3 <- anti_join(duplicate.hcris2, unique.hcris3, by=c("provider_number", "fyear"))
duplicate.hcris3 <- 
  duplicate.hcris3 %>%
  mutate(total_days=as.integer(total_days), time_diff=as.integer(time_diff)) %>%
  mutate_at(c("tot_charges","tot_discounts", "tot_operating_exp", "ip_charges",
              "icu_charges", "ancillary_charges", "tot_discharges", "mcare_discharges",
              "mcaid_discharges", "tot_mcare_payment", "secondary_mcare_payment",
              "hvbp_payment", "hrrp_payment", "uncomp_care", "tot_uncomp_care_charges", "bad_debt",
              "tot_uncomp_care_partial_pmts", "new_cap_ass", "cash", "net_pat_rev", "fixed_assets","accum_dep",
              "current_assets","current_liabilities"),
            list(~ .*(time_diff/total_days)))

## form weighted average of values for each fiscal year
unique.hcris4 <- 
  duplicate.hcris3 %>%
  group_by(provider_number, fyear) %>%
  mutate(hrrp_payment=if_else(is.na(hrrp_payment),0,hrrp_payment),
         hvbp_payment=if_else(is.na(hvbp_payment),0,hvbp_payment)) %>%
  summarize(beds=max(beds), tot_charges=sum(tot_charges), tot_discounts=sum(tot_discounts),
            tot_operating_exp=sum(tot_operating_exp), ip_charges=sum(ip_charges),
            icu_charges=sum(icu_charges), ancillary_charges=sum(ancillary_charges),
            tot_discharges=sum(tot_discharges), mcare_discharges=sum(mcare_discharges),
            mcaid_discharges=sum(mcaid_discharges), tot_mcare_payment=sum(tot_mcare_payment),
            secondary_mcare_payment=sum(secondary_mcare_payment), hvbp_payment=sum(hvbp_payment),
            hrrp_payment=sum(hrrp_payment), fy_start=min(fy_start), fy_end=max(fy_end),
            date_processed=max(date_processed), date_created=min(date_created), 
            street=first(street), city=first(city), state=first(state),
            zip=first(zip), county=first(county), name=first(name),
            uncomp_care = sum(uncomp_care), tot_uncomp_care_charges=sum(tot_uncomp_care_charges),
            tot_uncomp_care_partial_pmts=sum(tot_uncomp_care_partial_pmts), 
            bad_debt=sum(bad_debt), cost_to_charge=first(cost_to_charge),
            new_cap_ass=sum(new_cap_ass),
            cash=sum(cash),
            net_pat_rev=sum(net_pat_rev),
            fixed_assets=sum(fixed_assets), accum_dep=sum(accum_dep),
            current_assets=sum(current_assets), current_liabilities=sum(current_liabilities)) %>%
  mutate(source='weighted_average')

  


# Save final data ---------------------------------------------------------

final.hcris.data <- rbind(unique.hcris1, unique.hcris2, unique.hcris3, unique.hcris4)
final.hcris.data <- 
  final.hcris.data %>%
  rename(year=fyear) %>%
  arrange(provider_number, year)

write_tsv(final.hcris.data,'data/output/HCRIS_Data.txt')

File: data-code-py/_HCRIS_Data.py
"""HCRIS pipeline orchestrator.

Sources H3 (PPS 1985-1999), H1 (v1996 1998-2011), H2 (v2010 2010-2025),
adds missing variables, deduplicates overlapping reports, and writes the
final combined dataset.
"""

import numpy as np
import pandas as pd

from H3_HCRIS_PPS import extract_pps
from H1_HCRISv1996 import extract_v1996
from H2_HCRISv2010 import extract_v2010


# NA-safe aggregation helpers ----------------------------------------------

def na_sum(x):
    return np.nan if x.isna().all() else x.sum(skipna=True)

def na_max(x):
    return x.iloc[0] if x.isna().all() else x.max(skipna=True)

def na_min(x):
    return x.iloc[0] if x.isna().all() else x.min(skipna=True)


# Depreciation columns ----------------------------------------------------

DEPR_COLS = [
    "depr_land", "depr_bldg", "depr_lease", "depr_fixed_equip",
    "depr_auto", "depr_major_equip", "depr_minor_equip",
]

# depr_HIT only exists in v2010
DEPR_COLS_V2010 = DEPR_COLS + ["depr_HIT"]


def compute_accum_dep(df, depr_cols):
    """Sum absolute values of depreciation columns; all-NA → NA."""
    dep_vals = df[depr_cols].apply(pd.to_numeric, errors="coerce").abs()
    all_na = dep_vals.isna().all(axis=1)
    return dep_vals.sum(axis=1, skipna=True).where(~all_na, other=np.nan)


# Numeric flow/stock columns for aggregation -------------------------------

SUM_COLS = [
    "tot_charges", "tot_discounts", "tot_operating_exp", "ip_charges",
    "icu_charges", "ancillary_charges", "tot_discharges", "mcare_discharges",
    "mcaid_discharges", "tot_mcare_payment", "secondary_mcare_payment",
    "hvbp_payment", "hrrp_payment", "uncomp_care", "tot_uncomp_care_charges",
    "tot_uncomp_care_partial_pmts", "bad_debt", "new_cap_ass",
    "cash", "net_pat_rev", "fixed_assets", "accum_dep",
    "current_assets", "current_liabilities",
    "pps_ip_charges", "pps_op_charges", "pps_mcare_cost", "pps_pgm_cost",
]


def _agg_reports(grp):
    """Aggregate a group of duplicate reports into a single row."""
    row = {
        "beds": na_max(grp["beds"]),
        "fy_start": na_min(grp["fy_start"]),
        "fy_end": na_max(grp["fy_end"]),
        "date_processed": na_max(grp["date_processed"]),
        "date_created": na_min(grp["date_created"]),
        "street": grp["street"].iloc[0],
        "city": grp["city"].iloc[0],
        "state": grp["state"].iloc[0],
        "zip": grp["zip"].iloc[0],
        "county": grp["county"].iloc[0],
        "name": grp["name"].iloc[0],
        "cost_to_charge": grp["cost_to_charge"].iloc[0],
    }
    for c in SUM_COLS:
        row[c] = na_sum(grp[c])
    return pd.Series(row)


# Main pipeline ------------------------------------------------------------

def main():
    # Step 1: Extract from all three sources
    pps = extract_pps()
    v1996 = extract_v1996()
    v2010 = extract_v2010()

    # Step 2: Harmonize columns across sources

    ## PPS: add v2010-only columns, compute accum_dep, drop depr_ columns
    for col in ["hvbp_payment", "hrrp_payment", "tot_uncomp_care_charges",
                "tot_uncomp_care_partial_pmts", "bad_debt"]:
        pps[col] = np.nan
    pps["accum_dep"] = compute_accum_dep(pps, DEPR_COLS)
    pps.drop(columns=DEPR_COLS, inplace=True)

    ## v1996: same treatment
    for col in ["hvbp_payment", "hrrp_payment", "tot_uncomp_care_charges",
                "tot_uncomp_care_partial_pmts", "bad_debt"]:
        v1996[col] = np.nan
    v1996["accum_dep"] = compute_accum_dep(v1996, DEPR_COLS)
    v1996.drop(columns=DEPR_COLS, inplace=True)

    ## v2010: add v1996-only columns, compute uncomp_care and accum_dep
    for col in ["pps_ip_charges", "pps_op_charges", "pps_mcare_cost", "pps_pgm_cost"]:
        v2010[col] = np.nan
    v2010["uncomp_care"] = (
        pd.to_numeric(v2010["tot_uncomp_care_charges"], errors="coerce")
        - pd.to_numeric(v2010["tot_uncomp_care_partial_pmts"], errors="coerce")
        + pd.to_numeric(v2010["bad_debt"], errors="coerce")
    )
    v2010["accum_dep"] = compute_accum_dep(v2010, DEPR_COLS_V2010)
    v2010.drop(columns=DEPR_COLS_V2010, inplace=True, errors="ignore")

    # Step 3: Combine and parse dates
    print("Combining sources and deduplicating...")
    combined = pd.concat([pps, v1996, v2010], ignore_index=True)

    for col in ["fy_start", "fy_end", "date_processed", "date_created"]:
        combined[col] = pd.to_datetime(combined[col], format="mixed", errors="coerce")

    combined["tot_discounts"] = combined["tot_discounts"].abs()
    combined["hrrp_payment"] = combined["hrrp_payment"].abs()

    # Ensure numeric types for aggregation columns
    for c in SUM_COLS + ["beds", "cost_to_charge"]:
        if c in combined.columns:
            combined[c] = pd.to_numeric(combined[c], errors="coerce")

    # Step 4: Source-priority deduplication (v2010 > v1996 > pps)
    priority_map = {"v2010": 3, "v1996": 2, "pps": 1}
    combined["source_priority"] = combined["data_source"].map(priority_map).fillna(0).astype(int)
    combined = (
        combined
        .sort_values(["provider_number", "fy_start", "fy_end",
                       "source_priority", "date_created"],
                      ascending=[True, True, True, False, False])
        .groupby(["provider_number", "fy_start", "fy_end"], dropna=False)
        .first()
        .reset_index()
    )
    combined.drop(columns=["source_priority"], inplace=True)

    # Derive fiscal year from fy_end
    combined["fyear"] = combined["fy_end"].dt.year
    combined.drop(columns=["year"], inplace=True, errors="ignore")
    combined = combined.sort_values(["provider_number", "fyear"]).reset_index(drop=True)

    # Step 5: Count reports per provider/fyear
    combined["total_reports"] = combined.groupby(["provider_number", "fyear"])["provider_number"].transform("count")
    combined["report_number"] = combined.groupby(["provider_number", "fyear"]).cumcount() + 1

    # Step 6: Unique reports (one per provider/fyear)
    unique1 = (
        combined
        .loc[combined["total_reports"] == 1]
        .drop(columns=["report", "total_reports", "report_number", "npi", "status"])
        .assign(source="unique reports")
    )

    # Step 7: Duplicate reports
    dupes = combined.loc[combined["total_reports"] > 1].copy()
    dupes["time_diff"] = (dupes["fy_end"] - dupes["fy_start"]).dt.days

    grp_days = dupes.groupby(["provider_number", "fyear"])["time_diff"].transform("sum")
    dupes["total_days"] = grp_days

    ## 7a: Sum if total days < 370
    short_dupes = dupes.loc[dupes["total_days"] < 370].copy()
    short_dupes["hrrp_payment"] = short_dupes["hrrp_payment"].fillna(0)
    short_dupes["hvbp_payment"] = short_dupes["hvbp_payment"].fillna(0)

    if len(short_dupes) > 0:
        unique2 = (
            short_dupes
            .groupby(["provider_number", "fyear"])
            .apply(_agg_reports, include_groups=False)
            .reset_index()
            .assign(source="total for year")
        )
    else:
        unique2 = pd.DataFrame()

    ## 7b: Pick primary report if one covers full year
    long_dupes = dupes.loc[dupes["total_days"] >= 370].copy()
    long_dupes["max_days"] = long_dupes.groupby(["provider_number", "fyear"])["time_diff"].transform("max")
    long_dupes["max_date"] = long_dupes.groupby(["provider_number", "fyear"])["fy_end"].transform("max")

    primary_mask = (
        (long_dupes["max_days"] == long_dupes["time_diff"]) &
        (long_dupes["time_diff"] > 360) &
        (long_dupes["max_date"] == long_dupes["fy_end"])
    )
    unique3 = (
        long_dupes
        .loc[primary_mask]
        .drop(columns=["report", "total_reports", "report_number", "npi", "status",
                        "max_days", "time_diff", "total_days", "max_date"])
        .assign(source="primary report")
    )

    ## 7c: Weighted average for remainder
    used_keys = set()
    if len(unique3) > 0:
        used_keys = set(zip(unique3["provider_number"], unique3["fyear"]))

    remainder = long_dupes.loc[
        ~long_dupes.apply(lambda r: (r["provider_number"], r["fyear"]) in used_keys, axis=1)
    ].copy()

    if len(remainder) > 0:
        for c in SUM_COLS:
            remainder[c] = remainder[c] * (remainder["time_diff"] / remainder["total_days"])

        remainder["hrrp_payment"] = remainder["hrrp_payment"].fillna(0)
        remainder["hvbp_payment"] = remainder["hvbp_payment"].fillna(0)

        unique4 = (
            remainder
            .groupby(["provider_number", "fyear"])
            .apply(_agg_reports, include_groups=False)
            .reset_index()
            .assign(source="weighted_average")
        )
    else:
        unique4 = pd.DataFrame()

    # Step 8: Combine all tiers
    final = pd.concat([unique1, unique2, unique3, unique4], ignore_index=True)
    final = final.rename(columns={"fyear": "year"}).sort_values(["provider_number", "year"]).reset_index(drop=True)

    # Write output
    final.to_csv("data/output/HCRIS_Data.txt", sep="\t", index=False)
    print(f"Final dataset: {len(final):,} rows, {len(final.columns)} columns")
    print(f"Years: {int(final['year'].min())}-{int(final['year'].max())}")
    return final


if __name__ == "__main__":
    main()
